service/src/test/java/com/epam/esm/service/GiftCertificateServiceTest.java

```diff
@@ -0,0 +1,263 @@
+package com.epam.esm.service;
+
+import com.epam.esm.dto.GiftCertificateDto;
+import com.epam.esm.dto.GiftCertificateSearchParamsDto;
+import com.epam.esm.entity.GiftCertificate;
+import com.epam.esm.entity.Tag;
+import com.epam.esm.exception.EntityNotFoundException;
+import com.epam.esm.exception.InvalidEntityException;
+import com.epam.esm.repository.GiftCertificateRepository;
+import com.epam.esm.repository.OrderingType;
+import com.epam.esm.repository.TagRepository;
+import com.epam.esm.validator.GiftCertificateValidator;
+import com.epam.esm.validator.TagValidator;
+import org.junit.jupiter.api.*;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.*;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import java.math.BigDecimal;
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.TestInstance.Lifecycle.PER_CLASS;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+@TestInstance(PER_CLASS)
+class GiftCertificateServiceTest {
+    private static final ZonedDateTime INITIAL_DATE = LocalDateTime.now().atZone(ZoneOffset.UTC);
+
+    @InjectMocks
+    private GiftCertificateService certificateService;
+
+    @Mock
+    private GiftCertificateRepository certificateRepository;
+
+    @Mock
+    private TagRepository tagRepository;
+
+    @Spy
+    private GiftCertificateValidator certificateValidator;
+
+    @Spy
+    private TagValidator tagValidator;
```

Comment:
Why are `GiftCertificateValidator` and `TagValidator` spies? Why not mocks?

Reply:
Actually my tests need real validators rather than mocked. They are spies for verifying purposes.

Reply:
Fixed.